### ISSUE TITLE
Allow simultaneous computation of f, f'

### ISSUE THREAD

**Issue by cossio** [JuliaMath/Roots.jl#143]
Currently `find_zero` takes the function `f` and its derivative as separate arguments,

```
find_zero((f, df), ...)
```

It would be nice if one could instead call it as:

```
find_zero(fdf, ...)
```

where `fdf` is a *single* function that returns both `f(x)` and `f'(x)`. This can be very efficient if there are common computations involved in evaluating `f(x)` and `f'(x)`.

Or perhaps more generically,

```
find_zero(f, df, fdf, ...)
```

so that the algorithm can decide whether it needs only `f(x)`, only `f'(x)`, or both `f(x),f'(x)` (though in most cases I know, `fdf` suffices).

**Comment by jverzani** [comment]
There are only a few algorithms implemented that use this information (newton, halley, and an Order5 method). It could be implemented, but getting the signature right is a bit tricky. I'd like to keep the first argument as the function or collection of functions, so there would need to be some mechanism (a wrapping type?) to signal the object passed in will return.

This isn't so hard to incorporate. There is a FirstDerivative type that wraps the tuple so that calling the derivatives is easy. This could be generalized for other functions.

But I'm not sure it is worth the trouble, as it isn't clear these methods have any overall advantages. In a private branch, I'm reorganizing the tests, and I see this for a suite of functions over a range of starting points:

```
Method         fncnt exact fail maxresidual
Order0()       3874  114   23   1.7923440509548527e-6
Secant()       3677  80    27   1.4210854715202004e-7
Steffensen()   4507  79    40   1.8189894035458565e-12
Order5()       5722  93    33   3.4656721936698887e-6
Order8()       3660  100   36   1.7923440509548527e-6
Order16()      4291  106   34   9.947598300641403e-8
Newton         4984  92    33   1.5987211554602254e-8
Order5         6485  105   33   1.8189894035458565e-12
Halley         4617  94    30   1.2434497875801753e-8
```

The  methods with derivatives fail as often. Your plan could reduce the function calls, but they would still be less robust.

But let me see, if it isn't hard to mock up, I can take a look.

**Comment by cossio** [comment]
Amazing that `Order0()` has *less* function calls than Newton or Halley (I am assuming that's the `fncnt` column). What's the objective function? What is the meaning of the columns `exact`, `fail`?

**Comment by jverzani** [comment]
Newton, Halley and other have great convergence properties when started near the answer, but maybe not so when started farther away. (In fact, the fn calls counted here do not include the derivative.) These tests are testing some poor initial choices, hence the number of fails. Exact here isn't so impt, but it tells if the zero found has iszero(f(xstar)) or there is a sign change at xstar between adjacent floating point values. Also, the Order0 method switches to a typically fast bracketing method if a bracket is identified, and so that contributes relatively few function calls in these examples.

Anyways, it isn't too hard to test if there are speedups possible. It is just a question of how to expose the interface to the user. I'll have a look at the link you sent.

**Comment by jverzani** [comment]
I have a branch (https://github.com/jverzani/Roots.jl/tree/fdf) that prototypes this. (Which isn't quite right as convergence checks f(x_n-1) and not f(x_n), as needed.)

Here are some timings -- seems promising. This basically mirrors the speed up of using `sincos` over `(sin(x),cos(x))`:

```
julia> using BenchmarkTools

julia> @btime Roots.find_zero((sin,cos), 3.0, Roots.Newton())
  6.030 μs (83 allocations: 1.69 KiB)
3.141592653589793

julia> @btime Roots.find_zero(Roots.fg(sincos), 3.0, Roots.Newton())
  5.136 μs (80 allocations: 1.77 KiB)
3.141592653589793
```

The user would just wrap their function in `Roots.fg(...)` (or some such), as above.

**Comment by cossio** [comment]
Nice! For more complicated things the speedup can be more substantial. For example Lambert's W function has the property that evaluating `W(x)` is kind of costly, but once `W = W(x)` is known, the derivative comes for free, `W'(x) = W / (x * (1 + W))`. This would be a nice example.

**Comment by jverzani** [comment]
Give it a whirl and see if there are substantial savings. If so, I could try and polish this off.

**Comment by cossio** [comment]
Using https://github.com/cossio/LambertWExp.jl, I got:

```
julia> using Roots, BenchmarkTools, LambertWExp
julia> fdf(x) = (W = lambertwexp(x); (W - 1, W / (1+W)))
julia> f(x) = lambertwexp(x) - 1
julia> df(x) = (W = lambertwexp(x); W/(1+W))

julia> @btime find_zero(Roots.fg(fdf), 0, Roots.Newton())
  5.305 μs (154 allocations: 3.05 KiB)
1.0000000000000002

julia> @btime find_zero((f,df), 0, Roots.Newton())
  6.499 μs (168 allocations: 3.08 KiB)
1.0000000000000002

```

**Comment by jverzani** [comment]
Thanks. That's a good example. Surprisingly, Order1() will be faster by a significant amount here. Likely I have some type instability in the newton method that needs addressing.

**Comment by cossio** [comment]
Yes. Something's wrong. By the way I had to implement my own version of Newton's just to be able to pass `fdf`. My version is significantly faster than this, so I guess as you said there could be some type stability issue in `Roots.Newton`. 

```
julia> using LambertWExp, NewtonRaphsonSafe, BenchmarkTools
[ Info: Recompiling stale cache file /opt/julia-depot/compiled/v1.0/NewtonRaphsonSafe/XkKOu.ji for NewtonRaphsonSafe [495da384-cce8-11e8-2346-8f8bc5e73716]

julia> fdf(x) = (W = lambertwexp(x); (W - 1, W / (1+W)))
fdf (generic function with 1 method)

julia> @btime rtsafe(fdf, 0., -100, 100)
  724.015 ns (31 allocations: 496 bytes)
1.0000000000000002
```

In case it's helpful, you can find it here:

https://github.com/cossio/NewtonRaphsonSafe.jl

**Comment by jverzani** [comment]
Yes, the Roots functions have overhead and they are tuned so that the first call is faster (they don't specialize on the function argument). The `btime` macro won't show this, but `time` will. So custom methods come in faster in these tests. I have some "simple" functions for a few methods so that the package has faster equivalents, but nothing for Newton. Perhaps, adding this feature there is better.

**Comment by cossio** [comment]
You decided to go for it?

I don't understand why `f(x)/fp(x)` is needed for type stability. But still sometimes some factors cancel in the ratio and that's the thing that Newton actually needs, so it looks fine to me.

**Comment by jverzani** [comment]
Yes, its technical. Basically, for `find_zero` the state object store values of typeof x and typeof f. f' and f'' don't have that type necessarily, but both f/f' and f'/f'' have the units of x. I tried hiding that from the user, but in limited testing it seemed to slow things down. Maybe its a mistake, but it didn't seem so onerous, especially since it is there to provide a performance boost so some hoops to jump through are expected.

**Comment by cossio** [comment]
Cool. Thanks for this.
